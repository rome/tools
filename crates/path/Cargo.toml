[package]
name = "path"
version = "0.0.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
core = { path = "../core" }

[dev-dependencies]
