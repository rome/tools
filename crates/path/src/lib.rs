//! This module is responsible to manage paths inside Rome.
//! It is a small wrapper around [path::PathBuf] but it is also able to
//! give additional information around the the file that holds:
//! - the [FileHandlers] for the specific file
//! - shortcuts to open/write to the file
use core::{file_handlers::ExtensionHandler, App};
use std::{fs::File, io::Write, ops::Deref, path::PathBuf};

pub struct RomePath<'handler> {
	file: PathBuf,

	handler: Option<&'handler Box<dyn ExtensionHandler>>,
}

impl<'handler> Deref for RomePath<'handler> {
	type Target = PathBuf;

	fn deref(&self) -> &Self::Target {
		&self.file
	}
}

impl<'handler> RomePath<'handler> {
	pub fn new(path_to_file: &str) -> Self {
		Self {
			file: PathBuf::from(path_to_file),
			handler: None,
		}
	}

	/// Deduce the file handler based on the extension of the file.
	///
	/// Any error will default to the base file handler for now.
	///
	///
	/// ```rust
	/// use path::RomePath;
	/// use core::create_app;
	/// let app = create_app();
	/// let file = RomePath::new("file.js").deduce_handler(&app);
	/// let handler = file.get_handler();
	/// let expected = app.get_js_handler("js").unwrap();
	/// assert_eq!(handler.unwrap(), expected)
	/// ```
	pub fn deduce_handler(mut self, app: &'handler App) -> Self {
		if let None = self.extension() {
			return self;
		}
		let extension = self.extension().unwrap().to_str().unwrap();

		if let Some(handler) = app.get_handler(extension) {
			self.handler = Some(handler);
		}

		self
	}

	// TODO: handle error with diagnostic?
	/// Opens a file and returns a [File] in write mode
	pub fn open(&self) -> File {
		File::open(&self.file).expect("cannot open the file to format")
	}


	/// Accepts a file opened in read mode and saves into it
	pub fn save(&mut self, content: &str) -> Result<(), std::io::Error> {
		let mut file_to_write = File::create(&self.file).unwrap();
		file_to_write.write_all(content.as_bytes())
	}

	/// Returns the current handler associated to the file.
	///
	/// You need to call [deduce_handler] first in order to receive one. If not, [None] is always returned.
	pub fn get_handler(&self) -> Option<&Box<dyn ExtensionHandler>> {
		self.handler
	}
}

#[cfg(test)]
mod test {
	use crate::RomePath;
	use core::{
		create_app,
		file_handlers::{javascript::JsFileHandler, ExtensionHandler},
	};

	#[test]
	fn deduce_handler() {
		let app = create_app();
		let file = RomePath::new("file.js").deduce_handler(&app);
		let handler = file.get_handler();
		let expected = JsFileHandler {};
		assert_eq!(
			handler.unwrap().capabilities().format,
			expected.capabilities().format
		);
		assert_eq!(
			handler.unwrap().capabilities().lint,
			expected.capabilities().lint
		)
	}
}
