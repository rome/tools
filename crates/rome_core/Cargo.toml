[package]
name = "rome_core"
version = "0.0.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rome_fs = { version = "0.0.0", path = "../rome_fs" }
