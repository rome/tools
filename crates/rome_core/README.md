# `rome_core`

The crate contains the core functionality of Rome.

## Local installation

```toml
rome_core = { version = "0.0.0", path = "../rome_core" }
```
