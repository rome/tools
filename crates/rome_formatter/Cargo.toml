[package]
name = "rome_formatter"
version = "0.0.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rome_js_parser = { path = "../rome_js_parser" }
rome_js_syntax = { path = "../rome_js_syntax" }
rome_rowan = { path = "../rome_rowan" }
rome_fs = { path = "../rome_fs" }
rome_core = { path = "../rome_core" }
cfg-if = "1.0.0"
thiserror = "1.0.30"
tracing = "0.1.31"

[dev-dependencies]
serde = { version = "1", features = ["derive"] }
serde_json = "1"
tests_macros = { path = "../tests_macros" }
insta = { version = "1.10.0", features = ["glob"] }
rome_diagnostics = { path = "../rome_diagnostics" }
parking_lot = "0.12.0"
similar = "2.1.0"
similar-asserts = "1.2.0"
ctor = "0.1.21"
