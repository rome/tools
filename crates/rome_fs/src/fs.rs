use std::io;
use std::panic::RefUnwindSafe;
use std::path::{Path, PathBuf};
use std::sync::Arc;

use crate::{PathInterner, RomePath};
use rome_diagnostics::{
    location::FileId,
    {console, Advices, Diagnostic, LogCategory, Visit},
};

mod memory;
mod os;

pub use memory::{ErrorEntry, MemoryFileSystem};
pub use os::OsFileSystem;
use rome_diagnostics::Error;
pub const CONFIG_NAME: &str = "rome.json";

pub trait FileSystem: Send + Sync + RefUnwindSafe {
    /// It opens a file with the given set of options
    fn open_with_options(&self, path: &Path, options: OpenOptions) -> io::Result<Box<dyn File>>;

    /// Initiate a traversal of the filesystem
    ///
    /// This method creates a new "traversal scope" that can be used to
    /// efficiently batch many filesystem read operations
    fn traversal<'scope>(&'scope self, func: BoxedTraversal<'_, 'scope>);

    /// Returns the name of the main configuration file
    fn config_name(&self) -> &str {
        CONFIG_NAME
    }
}

pub trait File {
    /// Read the content of the file into `buffer`
    fn read_to_string(&mut self, buffer: &mut String) -> io::Result<()>;

    /// Overwrite the content of the file with the provided bytes
    ///
    /// This will write to the associated memory buffer, as well as flush the
    /// new content to the disk if this is a physical file
    fn set_content(&mut self, content: &[u8]) -> io::Result<()>;
}

/// This struct is a "mirror" of [std::fs::FileOptions].
/// Refer to their documentation for more details
#[derive(Default)]
pub struct OpenOptions {
    read: bool,
    write: bool,
    append: bool,
    truncate: bool,
    create: bool,
    create_new: bool,
}

impl OpenOptions {
    pub fn read(mut self, read: bool) -> Self {
        self.read = read;
        self
    }
    pub fn write(mut self, write: bool) -> Self {
        self.write = write;
        self
    }
    pub fn append(mut self, append: bool) -> Self {
        self.append = append;
        self
    }
    pub fn truncate(mut self, truncate: bool) -> Self {
        self.truncate = truncate;
        self
    }
    pub fn create(mut self, create: bool) -> Self {
        self.create = create;
        self
    }
    pub fn create_new(mut self, create_new: bool) -> Self {
        self.create_new = create_new;
        self
    }

    pub fn into_fs_options(self, options: &mut std::fs::OpenOptions) -> &mut std::fs::OpenOptions {
        options
            .read(self.read)
            .write(self.write)
            .append(self.append)
            .truncate(self.truncate)
            .create(self.create)
            .create_new(self.create_new)
    }
}

/// Trait that contains additional methods to work with [FileSystem]
pub trait FileSystemExt: FileSystem {
    /// Open a file with `read` and `write` options
    fn open(&self, path: &Path) -> io::Result<Box<dyn File>> {
        self.open_with_options(path, OpenOptions::default().read(true).write(true))
    }
    /// Open a file with `write` and `create_new` options
    fn create(&self, path: &Path) -> io::Result<Box<dyn File>> {
        self.open_with_options(path, OpenOptions::default().write(true).create_new(true))
    }
    /// Opens a file with read options
    fn read(&self, path: &Path) -> io::Result<Box<dyn File>> {
        self.open_with_options(path, OpenOptions::default().read(true).create_new(true))
    }
}

type BoxedTraversal<'fs, 'scope> = Box<dyn FnOnce(&dyn TraversalScope<'scope>) + Send + 'fs>;

pub trait TraversalScope<'scope> {
    /// Spawn a new filesystem read task
    ///
    /// If the provided path exists and is a file, then the [`handle_file`](TraversalContext::handle_file)
    /// method of the provided [TraversalContext] will be called. If it's a
    /// directory, it will be recursively traversed and all the files the
    /// [`can_handle`](TraversalContext::can_handle) method of the context
    /// returns true for will be handled as well
    fn spawn(&self, context: &'scope dyn TraversalContext, path: PathBuf);
}

pub trait TraversalContext: Sync {
    /// Provides the traversal scope with an instance of [PathInterner], used
    /// to emit diagnostics for IO errors that may happen in the traversal process
    fn interner(&self) -> &PathInterner;

    /// Called by the traversal process to emit an error diagnostic associated
    /// with a particular file ID when an IO error happens
    fn push_diagnostic(&self, error: Error);

    /// Checks if the traversal context can handle a particular path, used as
    /// an optimization to bail out of scheduling a file handler if it wouldn't
    /// be able to process the file anyway
    fn can_handle(&self, path: &RomePath) -> bool;

    /// This method will be called by the traversal for each file it finds
    /// where [TraversalContext::can_handle] returned true
    fn handle_file(&self, path: &Path, file_id: FileId);
}

impl<T> FileSystem for Arc<T>
where
    T: FileSystem + Send,
{
    fn open_with_options(&self, path: &Path, options: OpenOptions) -> io::Result<Box<dyn File>> {
        T::open_with_options(self, path, options)
    }

    fn traversal<'scope>(&'scope self, func: BoxedTraversal<'_, 'scope>) {
        T::traversal(self, func)
    }
}

#[derive(Debug, Diagnostic)]
#[diagnostic(severity = Warning, category = "internalError/fs")]
struct FileSystemDiagnostic {
    #[location(resource)]
    path: String,
    #[message]
    #[description]
    #[advice]
    error_kind: ErrorKind,
}

#[derive(Clone, Debug)]
enum ErrorKind {
    /// Unknown file type
    UnknownFileType,
    /// Dereferenced (broken) symbolic link
    DereferencedSymlink(String),
    /// Symbolic link cycle or symbolic link infinite expansion
    InfiniteSymlinkExpansion(String),
}

impl console::fmt::Display for ErrorKind {
    fn fmt(&self, fmt: &mut console::fmt::Formatter) -> io::Result<()> {
        match self {
            ErrorKind::UnknownFileType => fmt.write_str("Unknown file type"),
            ErrorKind::DereferencedSymlink(_) => fmt.write_str("Dereferenced symlink"),
            ErrorKind::InfiniteSymlinkExpansion(_) => fmt.write_str("Infinite symlink expansion"),
        }
    }
}

impl std::fmt::Display for ErrorKind {
    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ErrorKind::UnknownFileType => write!(fmt, "Unknown file type"),
            ErrorKind::DereferencedSymlink(_) => write!(fmt, "Dereferenced symlink"),
            ErrorKind::InfiniteSymlinkExpansion(_) => write!(fmt, "Infinite symlink expansion"),
        }
    }
}

impl Advices for ErrorKind {
    fn record(&self, visitor: &mut dyn Visit) -> io::Result<()> {
        match self {
            ErrorKind::UnknownFileType => visitor.record_log(
                LogCategory::Info,
                &"Rome encountered a file system entry that's neither a file, directory or symbolic link",
            ),
            ErrorKind::DereferencedSymlink(path) => visitor.record_log(
                LogCategory::Info,
                &format!("Rome encountered a file system entry that is a broken symbolic link: {}", path),
            ),
            ErrorKind::InfiniteSymlinkExpansion(path) => visitor.record_log(
                LogCategory::Error,
                &format!("Rome encountered a file system entry that leads to an infinite symbolic link expansion, causing an infinite cycle: {}", path),
            ),
        }
    }
}
