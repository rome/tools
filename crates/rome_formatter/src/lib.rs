//! Rome's official formatter.
//!
//! The crate exposes some API and utilities to implement the formatting logic.
//!
//! The formatter relies on an [IR], which allows to format any kind of data structure.
//!
//! In order to implement the formatting logic, you need to implement the trait [FormatValue] for
//! the data structure you want to format.
//!
//! Let's say, for example that you have a small data structure that represents a key/value data:
//!
//! ```rust,no_test
//! struct KeyValue {
//!     key: &'static str,
//!     value: &'static str
//! }
//! ```
//!
//! Now, we do want to create this IR for the data structure:
//! ```rust
//! use rome_formatter::{format_elements, format_element, Formatter, ToFormatElement, FormatElement, FormatResult, FormatOptions, space_token, token };
//!
//! struct KeyValue {
//!     key: &'static str,
//!     value: &'static str
//! }
//!
//! impl ToFormatElement for KeyValue {
//!     fn to_format_element(&self, formatter: &Formatter)-> FormatResult<FormatElement>  {
//!         Ok(format_elements![
//!             token(self.key),
//!             space_token(),
//!             token("=>"),
//!             space_token(),
//!             token(self.value)
//!         ])
//!     }
//! }
//!
//! fn my_function() {
//!     let key_value = KeyValue { key: "lorem", value: "ipsum" };
//!     let element = key_value.to_format_element(&Formatter::default()).unwrap();
//!     let result = format_element(&element, FormatOptions::default());
//!     assert_eq!(result.as_code(), "lorem => ipsum");
//! }
//!
//! ```
//! [IR]: https://en.wikipedia.org/wiki/Intermediate_representation

mod cst;
mod format_element;
mod format_elements;
mod formatter;
pub mod formatter_traits;
mod intersperse;
mod js;
mod printer;
mod ts;
mod utils;

pub use formatter::Formatter;
use rome_rowan::TextRange;
use rome_rowan::TextSize;
use rome_rowan::TokenAtOffset;
use rslint_parser::{parse, SourceType, SyntaxError, SyntaxNode};
use std::fmt::Display;

pub use format_element::{
    block_indent, concat_elements, empty_element, empty_line, fill_elements, group_elements,
    hard_group_elements, hard_line_break, if_group_breaks, if_group_fits_on_single_line, indent,
    join_elements, join_elements_hard_line, line_suffix, soft_block_indent, soft_line_break,
    soft_line_break_or_space, soft_line_indent_or_space, space_token, token, FormatElement, Token,
};
pub use printer::Printer;
pub use printer::PrinterOptions;
use rome_core::App;
use rome_path::RomePath;
use std::str::FromStr;
use thiserror::Error;

/// This trait should be implemented on each node/value that should have a formatted representation
pub trait ToFormatElement {
    fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement>;
}

/// Public return type of the formatter
pub type FormatResult<F> = Result<F, FormatError>;

#[derive(Debug, PartialEq, Error)]
/// Series of errors encountered during formatting
pub enum FormatError {
    /// Node is missing and it should be required for a correct formatting
    #[error("missing required child")]
    MissingRequiredChild,

    /// In case our formatter doesn't know how to format a certain language
    #[error("language is not supported")]
    UnsupportedLanguage,

    /// When the ability to format the current file has been turned off on purpose
    #[error("formatting capability is disabled")]
    CapabilityDisabled,
}

impl From<SyntaxError> for FormatError {
    fn from(syntax_error: SyntaxError) -> Self {
        match syntax_error {
            SyntaxError::MissingRequiredChild(_node) => FormatError::MissingRequiredChild,
        }
    }
}

impl From<&SyntaxError> for FormatError {
    fn from(syntax_error: &SyntaxError) -> Self {
        match syntax_error {
            SyntaxError::MissingRequiredChild(_node) => FormatError::MissingRequiredChild,
        }
    }
}

#[derive(Debug, Eq, PartialEq, Clone, Copy)]
pub enum IndentStyle {
    /// Tab
    Tab,
    /// Space, with its quantity
    Space(u8),
}

impl Default for IndentStyle {
    fn default() -> Self {
        Self::Tab
    }
}

impl FromStr for IndentStyle {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "tab" => Ok(Self::Tab),
            "space" => Ok(Self::Space(2)),
            // TODO: replace this error with a diagnostic
            _ => Err("Value not supported for IndentStyle"),
        }
    }
}

impl Display for IndentStyle {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            IndentStyle::Tab => write!(f, "Tab"),
            IndentStyle::Space(size) => write!(f, "Spaces, size: {}", size),
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct FormatOptions {
    /// The indent style
    pub indent_style: IndentStyle,

    /// What's the max width of a line. Defaults to 80
    pub line_width: u16,
}

impl FormatOptions {
    pub fn new(indent_style: IndentStyle) -> Self {
        Self {
            indent_style,
            ..Self::default()
        }
    }
}

impl Default for FormatOptions {
    fn default() -> Self {
        Self {
            indent_style: IndentStyle::default(),
            line_width: 80,
        }
    }
}

impl Display for FormatOptions {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Indent style: {}", self.indent_style)?;
        writeln!(f, "Line width: {}", self.line_width)?;
        Ok(())
    }
}

/// Lightweight sourcemap marker between source and output tokens
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct SourceMarker {
    /// Position of the marker in the original source
    pub source: TextSize,
    /// Position of the marker in the output code
    pub dest: TextSize,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Formatted {
    code: String,
    range: Option<TextRange>,
    sourcemap: Vec<SourceMarker>,
    verbatim_source: Vec<(String, TextRange)>,
}

impl Formatted {
    fn new(
        code: String,
        range: Option<TextRange>,
        sourcemap: Vec<SourceMarker>,
        verbatim_source: Vec<(String, TextRange)>,
    ) -> Self {
        Self {
            code,
            range,
            sourcemap,
            verbatim_source,
        }
    }

    /// Construct an empty formatter result
    fn new_empty() -> Self {
        Self {
            code: String::new(),
            range: None,
            sourcemap: Vec::new(),
            verbatim_source: Vec::new(),
        }
    }

    /// Range of the input source file covered by this formatted code,
    /// or None if the entire file is covered in this instance
    pub fn range(&self) -> Option<TextRange> {
        self.range
    }

    /// Returns a list of [SourceMarker] mapping byte positions
    /// in the output string to the input source code
    pub fn sourcemap(&self) -> &[SourceMarker] {
        &self.sourcemap
    }

    /// Access the resulting code, borrowing the result
    pub fn as_code(&self) -> &str {
        &self.code
    }

    /// Access the resulting code, consuming the result
    pub fn into_code(self) -> String {
        self.code
    }

    pub fn verbatim(&self) -> &[(String, TextRange)] {
        &self.verbatim_source
    }
}

/// Formats a JavaScript (and its super languages) file based on its features.
///
/// It returns a [Formatted] result, which the user can use to override a file.
pub fn format(options: FormatOptions, syntax: &SyntaxNode) -> FormatResult<Formatted> {
    let element = Formatter::new(options).format_root(syntax)?;
    Ok(Printer::new(options).print(&element))
}

/// Outputs formatter IR for a JavaScript (and its super languages) file
///
/// It returns a [FormatElement] result. Mostly for debugging purposes.
pub fn to_format_element(
    options: FormatOptions,
    syntax: &SyntaxNode,
) -> FormatResult<FormatElement> {
    Formatter::new(options).format_root(syntax)
}

/// Formats a range within a file, supported by Rome
///
/// This runs a simple heuristic to determine the initial indentation
/// level of the node based on the provided [FormatOptions], which
/// must match currently the current initial of the file. Additionally,
/// because the reformatting happens only locally the resulting code
/// will be indented with the same level as the original selection,
/// even if it's a mismatch from the rest of the block the selection is in
///
/// It returns a [Formatted] result with a range corresponding to the
/// range of the input that was effectively overwritten by the formatter
pub fn format_range(
    options: FormatOptions,
    root: &SyntaxNode,
    range: TextRange,
) -> FormatResult<Formatted> {
    // Find the tokens corresponding to the start and end of the range
    let start_token = root.token_at_offset(range.start());
    let end_token = root.token_at_offset(range.end());

    // If these tokens were not found this means either:
    // 1. The input [SyntaxNode] was empty
    // 2. The input node was not the root [SyntaxNode] of the file
    // In the first case we can return an empty result immediately,
    // otherwise default to the first and last tokens in the root node
    let start_token = match start_token {
        // If the start of the range lies between two tokens,
        // start at the rightmost one
        TokenAtOffset::Between(_, token) => token,
        TokenAtOffset::Single(token) => token,
        TokenAtOffset::None => match root.first_token() {
            Some(token) => token,
            // root node is empty
            None => return Ok(Formatted::new_empty()),
        },
    };
    let end_token = match end_token {
        // If the end of the range lies between two tokens,
        // end at the leftmost one
        TokenAtOffset::Between(token, _) => token,
        TokenAtOffset::Single(token) => token,
        TokenAtOffset::None => match root.last_token() {
            Some(token) => token,
            // root node is empty
            None => return Ok(Formatted::new_empty()),
        },
    };

    // Find the lowest common ancestor node for the start and end token
    // by building the path to the root node from both tokens and
    // iterating along the two paths at once to find the first divergence
    #[allow(clippy::needless_collect)]
    let start_to_root: Vec<_> = start_token.ancestors().collect();
    #[allow(clippy::needless_collect)]
    let end_to_root: Vec<_> = end_token.ancestors().collect();

    let common_root = start_to_root
        .into_iter()
        .rev()
        .zip(end_to_root.into_iter().rev())
        .map_while(|(lhs, rhs)| if lhs == rhs { Some(lhs) } else { None })
        .last();

    // Logically this should always return at least the root node,
    // fallback to said node just in case
    let common_root = common_root.as_ref().unwrap_or(root);

    // Perform the actual formatting of the root node with
    // an appropriate indentation level
    let formatted = format_node(options, common_root)?;

    // This finds the closest marker to the beginning of the source
    // starting before or at said starting point, and the closest
    // marker to the end of the source range starting after or at
    // said ending point respectively
    let mut range_start = None;
    let mut range_end = None;

    for marker in &formatted.sourcemap {
        if let Some(start_dist) = marker.source.checked_sub(range.start()) {
            range_start = match range_start {
                Some((prev_marker, prev_dist)) => {
                    if start_dist < prev_dist {
                        Some((marker, start_dist))
                    } else {
                        Some((prev_marker, prev_dist))
                    }
                }
                None => Some((marker, start_dist)),
            }
        }

        if let Some(end_dist) = range.end().checked_sub(marker.source) {
            range_end = match range_end {
                Some((prev_marker, prev_dist)) => {
                    if end_dist < prev_dist {
                        Some((marker, end_dist))
                    } else {
                        Some((prev_marker, prev_dist))
                    }
                }
                None => Some((marker, end_dist)),
            }
        }
    }

    // If no start or end were found, this means that the edge of the formatting
    // range was near the edge of the input, and no marker were emitted before
    // the start (or after the end) of the formatting range: in this case
    // the start/end marker default to the start/end of the input
    let (start_source, start_dest) = match range_start {
        Some((start_marker, _)) => (start_marker.source, start_marker.dest),
        None => (common_root.text_range().start(), TextSize::from(0)),
    };
    let (end_source, end_dest) = match range_end {
        Some((end_marker, _)) => (end_marker.source, end_marker.dest),
        None => (
            common_root.text_range().end(),
            TextSize::try_from(formatted.code.len()).expect("code length out of bounds"),
        ),
    };

    let input_range = TextRange::new(start_source, end_source);
    let output_range = TextRange::new(start_dest, end_dest);
    let code = &formatted.code[output_range];

    Ok(Formatted::new(
        code.into(),
        Some(input_range),
        formatted.sourcemap,
        formatted.verbatim_source,
    ))
}

/// Formats a single node within a file, supported by Rome
///
/// This runs a simple heuristic to determine the initial indentation
/// level of the node based on the provided [FormatOptions], which
/// must match currently the current initial of the file. Additionally,
/// because the reformatting happens only locally the resulting code
/// will be indented with the same level as the original selection,
/// even if it's a mismatch from the rest of the block the selection is in
///
/// It returns a [Formatted] result
pub fn format_node(options: FormatOptions, root: &SyntaxNode) -> FormatResult<Formatted> {
    // Determine the initial indentation level for the printer by inspecting the trivias
    // of each token from the first token of the common root towards the start of the file
    let mut tokens = std::iter::successors(root.first_token(), |token| token.prev_token());

    // From the iterator of tokens, build an iterator of trivia pieces (once again the iterator is
    // reversed, starting from the last trailing trivia towards the first leading trivia).
    // The first token is handled specially as we only wan to consider its leading trivias
    let first_token = tokens.next();
    let first_token_trivias = first_token
        .into_iter()
        .flat_map(|token| token.leading_trivia().pieces().rev());

    let next_tokens_trivias = tokens.flat_map(|token| {
        token
            .trailing_trivia()
            .pieces()
            .rev()
            .chain(token.leading_trivia().pieces().rev())
    });

    let trivias = first_token_trivias
        .chain(next_tokens_trivias)
        .filter(|piece| {
            // We're only interested in newline and whitespace trivias, skip over comments
            let is_newline = piece.is_newline();
            let is_whitespace = piece.is_whitespace();
            is_newline || is_whitespace
        });

    // Finally run the iterator until a newline trivia is found, and get the last whitespace trivia before it
    let last_whitespace = trivias.map_while(|piece| piece.as_whitespace()).last();
    let initial_indent = match last_whitespace {
        Some(trivia) => {
            // This logic is based on the formatting options passed in
            // the be user (or the editor) as we do not have any kind
            // of indentation type detection yet. Unfortunately this
            // may not actually match the current content of the file
            let length = trivia.text().len() as u16;
            match options.indent_style {
                IndentStyle::Tab => length,
                IndentStyle::Space(width) => length / u16::from(width),
            }
        }
        // No whitespace was found between the start of the range
        // and the start of the file
        None => 0,
    };

    let element = Formatter::new(options).format_root(root)?;
    let formatted = Printer::new(options).print_with_indent(&element, initial_indent);

    Ok(Formatted::new(
        formatted.code,
        Some(root.text_range()),
        formatted.sourcemap,
        formatted.verbatim_source,
    ))
}

pub fn format_element(element: &FormatElement, options: FormatOptions) -> Formatted {
    let printer = Printer::new(options);
    printer.print(element)
}

pub fn format_file_and_save(rome_path: &mut RomePath, options: FormatOptions, app: &App) {
    let result = if app.can_format(rome_path) {
        let buffer = rome_path.get_buffer_from_file();
        let source_type = rome_path
            .as_path()
            .try_into()
            .unwrap_or_else(|_| SourceType::js_module());
        let root = parse(buffer.as_str(), 0, source_type).syntax();
        Some(format(options, &root))
    } else {
        None
    };
    if let Some(Ok(result)) = result {
        rome_path
            .save(result.as_code())
            .expect("Could not write the formatted code on file");
    }
}

#[cfg(test)]
mod tests {

    use super::{format_range, FormatOptions};
    use crate::IndentStyle;
    use rome_rowan::{TextRange, TextSize};
    use rslint_parser::parse_script;

    #[test]
    fn test_range_formatting() {
        let input = "
while(
    true
) {
    function func() {
    func(     /* comment */
    );

    let array =
        [ 1
    , 2];

    }

    function func2()
    {

    const no_format    =    () => {};

    }
}
";

        // Start the formatting range two characters before the "let" keywords,
        // in the middle of the indentation whitespace for the line
        let range_start = TextSize::try_from(input.find("let").unwrap() - 2).unwrap();
        let range_end = TextSize::try_from(input.find("const").unwrap()).unwrap();

        let tree = parse_script(input, 0);
        let result = format_range(
            FormatOptions {
                indent_style: IndentStyle::Space(4),
                ..FormatOptions::default()
            },
            &tree.syntax(),
            TextRange::new(range_start, range_end),
        );

        let result = result.expect("range formatting failed");
        assert_eq!(
            result.range(),
            Some(TextRange::new(range_start + TextSize::from(2), range_end))
        );
        assert_eq!(
            result.as_code(),
            "let array = [1, 2];\n    }\n\n    function func2() {\n        "
        );
    }

    #[test]
    fn test_range_formatting_indentation() {
        let input = "
function() {
         const veryLongIdentifierToCauseALineBreak = { veryLongKeyToCauseALineBreak: 'veryLongValueToCauseALineBreak' }
}
";

        let range_start = TextSize::try_from(input.find("const").unwrap()).unwrap();
        let range_end = TextSize::try_from(input.find('}').unwrap()).unwrap();

        let tree = parse_script(input, 0);
        let result = format_range(
            FormatOptions {
                indent_style: IndentStyle::Space(4),
                ..FormatOptions::default()
            },
            &tree.syntax(),
            TextRange::new(range_start, range_end),
        );

        let result = result.expect("range formatting failed");
        assert_eq!(result.range(), Some(TextRange::new(range_start, range_end)));
        // As a result of the indentation normalization, the number of spaces within
        // the object expression is currently rounded down from an odd indentation level
        assert_eq!(
            result.as_code(),
            "const veryLongIdentifierToCauseALineBreak = {\n            veryLongKeyToCauseALineBreak: \"veryLongValueToCauseALineBreak\",\n        "
        );
    }
}

#[cfg(test)]
mod test {
    use crate::format;
    use crate::FormatOptions;
    use rslint_parser::{parse, SourceType};

    #[test]
    #[ignore]
    // use this test check if your snippet prints as you wish, without using a snapshot
    fn quick_test() {
        let src = r#"
        `something ${ () => { var hey; const looooooooooong_expression = "loooooooooong_expression" }} something else ${ ehy }`;
"#;
        let syntax = SourceType::ts();
        let tree = parse(src, 0, syntax);
        let result = format(FormatOptions::default(), &tree.syntax()).unwrap();
        assert_eq!(
            result.as_code(),
            r#"let g = [[], [0, 1], [0, 1]];
"#
        );
    }
}
