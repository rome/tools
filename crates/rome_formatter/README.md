# `rome_formatter`

The official formatter used by Rome.

## Internal installation

```toml
rome_formatter = { version = "0.0.0", path = "../rome_formatter" }
```

## Usage

The foundation of the formatter relies on two pillars:

- the usage of a [*trait*](https://doc.rust-lang.org/reference/items/traits.html) called `ToFormatElement`;
- the creation of an intermediate IR via a series of helpers

Import the `ToFormatElement` trait and implement it for the data structure you need.

```rust
use rome_formatter::{ToFormatElement, FormatElement, format_elements, token}

struct Buzz {
 blast: String
}

impl ToFormatElement for Buzz {
 fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
 // implementation goes here
 format_elements![token("_"), blast.as_str(), token("_")]
 }
}

```

## Rules when formatting AST nodes

1. if a token is mandatory and the AST has that information, please use that token instead, for example:

	```rust
	fn to_format_element(&self, formatter: &Formatter) -> FormatResult<FormatElement> {
		let l_paren_yes = formatter.format_token(&self.l_paren_token()?)?; // yes
		let l_paren_no = toke("("); // no
	}
	```

 1. for tokens that are not mandatory, use our helpers
 1. do not attempt to "fix" the code. If you know a token/node is mandatory, return `None` instead
