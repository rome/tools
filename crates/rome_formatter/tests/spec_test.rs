use rome_core::App;
use rome_formatter::{format, FormatOptions};
use rome_path::RomePath;
use rslint_parser::{parse, Syntax};
use std::fs;
use std::path::Path;

/// [insta.rs](https://insta.rs/docs) snapshot testing
///
/// For better development workflow, run
/// `cargo watch -i '*.new' -x 'test -p rome_formatter formatter'`
///
/// To review and commit the snapshots, `cargo install cargo-insta`, and run
/// `cargo insta review` or `cargo insta accept`
///
/// The input and the expected output are stored as dedicated files in the `tests/specs` directory where
/// the input file name is `{spec_name}.json` and the output file name is `{spec_name}.json.snap`.
///
/// Specs can be grouped in directories by specifying the directory name in the spec name. Examples:
///
/// # Examples
///
/// * `json/null` -> input: `tests/specs/json/null.json`, expected output: `tests/specs/json/null.json.snap`
/// * `null` -> input: `tests/specs/null.json`, expected output: `tests/specs/null.json.snap`
pub fn run(spec_input_file: &str, _: &str, file_type: &str) {
    let app = App::new();
    let file_path = &spec_input_file;
    let spec_input_file = Path::new(spec_input_file);

    assert!(
        spec_input_file.is_file(),
        "The input '{}' must exist and be a file.",
        spec_input_file.display()
    );

    let mut rome_path = RomePath::new(file_path);
    if app.can_format(&rome_path) {
        let buffer = rome_path.get_buffer_from_file();
        let syntax = if file_type == "module" {
            Syntax::default().module()
        } else {
            Syntax::default()
        };

        let root = parse(buffer.as_str(), 0, syntax).syntax();
        let formatted_result = format(FormatOptions::default(), &root);
        let file_name = spec_input_file.file_name().unwrap().to_str().unwrap();
        let input = fs::read_to_string(file_path).unwrap();
        let result = formatted_result.unwrap();
        // we ignore the error for now
        let snapshot = format!("# Input\n{}\n---\n# Output\n{}", input, result.code());

        insta::with_settings!({
            prepend_module_to_snapshot => false,
            snapshot_path => spec_input_file.parent().unwrap(),
        }, {
            insta::assert_snapshot!(file_name, snapshot, file_name);
        });
    }
}
