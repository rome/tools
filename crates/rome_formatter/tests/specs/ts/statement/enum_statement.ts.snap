---
source: crates/rome_formatter/tests/spec_test.rs
assertion_line: 163
expression: enum_statement.ts

---
# Input
enum    A   {}
enum B {a = "something",
b = "something", c = "something",
d = "something",
e = "something", f = "something",
}

const enum C {
    A, B, C,
    D
    ,
    F
}
=============================
# Outputs
## Output 1
-----
Indent style: Tab
Line width: 80
-----
enum A { }
enum B {
	a = "something",
	b = "something",
	c = "something",
	d = "something",
	e = "something",
	f = "something",
}

const enum C { A, B, C, D, F }

